fix: Disabled elements being clickable

TheoAnastasiadis opened the pull request:
<!-- Thanks for contributing! PLEASE...
- Read our contributing guidelines: https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md 
- Read our Code Review Checklist on coding standards and what needs to be done before a PR can be merged: https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md#Code-Review-Checklist
- Mark this PR as "Draft" if it is not ready for review.
- Make sure you set the correct base branch based on what packages you're changing: https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md#branches
-->

- Closes #28788

### Additional details
Click commands on children of disabled elements still produce `click` events.
For example the following;
```html
<button disabled>
  <span>Click Me!</span>
</button>
```
```javascript
cy.contains("Click Me!").click() // this triggers `click` events (at least on Chrome).
```
Even though the `button` element is disabled.

### How has the user experience changed?
Little

### PR Tasks
<!-- 
These tasks must be completed before a PR is merged.
If a task does not apply, write [na] instead of checking the box.
DO NOT DELETE the PR checklist.
-->

- [x] Have tests been added/updated?
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? <!-- Link to PR here -->
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

TheoAnastasiadis committed:
fixes #28788 -- mouse.ts logic
If mouseUp element or mouseDown element or commonAncestor element is :disabled, click event should be prevented

TheoAnastasiadis committed:
fixes #28788 -- e2e test
tests that no click events are registered when click happens on child of disabled element

TheoAnastasiadis committed:
fixes #28788 -- mouse.ts logic
If mouseUp element or mouseDown element have an "actually disabled" parent, the click event should not be registered

TheoAnastasiadis committed:
refactoring: minor name changes for readability

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/cypress-io/cypress?pullRequest=28807) <br/>All committers have signed the CLA.

jennifer-shehane commented:
@TheoAnastasiadis Thanks for the contribution. Please see this failing check for details about adding a changelog entry: https://github.com/cypress-io/cypress/actions/runs/7670111753/job/20905527788?pr=28807

TheoAnastasiadis committed:
fixes #24322 --added changelog entry

AtofStryker commented on packages/driver/cypress/e2e/commands/actions/click.cy.js:
```diff
@@ -546,6 +546,27 @@ describe('src/cy/commands/actions/click', () => {
+      btn.attr('disabled', true)
+      btn.append(span)
+
+      cy.get('span').click()
```
Currently this is clicking all the spans on the page, which is about 38 of them. I think we just want to click the span inside the button?
```suggestion
      cy.get('button:first span').click()
```

AtofStryker commented on packages/driver/src/cy/mouse.ts:
```diff
@@ -567,6 +567,21 @@ export const create = (state: StateFunc, keyboard: Keyboard, focused: IFocused,
 
+        // Only send click event if element is not disabled.
+        // First find an parent element that can actually be disabled
+        const canBeDisabled = (el) => {
+          const actuallyDisabledElements = ['button', 'input', 'select', 'textarea', 'optgroup', 'option', 'fieldset']
+
+          return actuallyDisabledElements.indexOf($elements.getTagName(el)) > -1
+        }
```
I had some issues getting this code to run. In the current state, `canBeDisabled` returns a `boolean` when `findParent` is expecting an element. the boolean is passed into `$elements.isDisabled` and is failing since it's expecting a JQuery selector. Were you able to get this to run OK? Either way I think we can get it working with the following suggestions!

```suggestion
        const findParentThatCanBeDisabled = (el: HTMLElement): HTMLElement | null => {
          const elementsThatCanBeDisabled = ['button', 'input', 'select', 'textarea', 'optgroup', 'option', 'fieldset']

          return elementsThatCanBeDisabled.includes($elements.getTagName(el)) ? el : null
        }
        }
```

AtofStryker commented on packages/driver/src/cy/mouse.ts:
```diff
@@ -567,6 +567,21 @@ export const create = (state: StateFunc, keyboard: Keyboard, focused: IFocused,
+          return actuallyDisabledElements.indexOf($elements.getTagName(el)) > -1
+        }
+
+        const parentThatCanBeDisabled = $elements.findParent(mouseUpPhase.targetEl, canBeDisabled) || $elements.findParent(mouseDownPhase.targetEl, canBeDisabled)
```
```suggestion
        const parentThatCanBeDisabled = $elements.findParent(mouseUpPhase.targetEl, findParentThatCanBeDisabled) || $elements.findParent(mouseDownPhase.targetEl, findParentThatCanBeDisabled)
```

AtofStryker commented on packages/driver/src/cy/mouse.ts:
```diff
@@ -567,6 +567,21 @@ export const create = (state: StateFunc, keyboard: Keyboard, focused: IFocused,
+        const parentThatCanBeDisabled = $elements.findParent(mouseUpPhase.targetEl, canBeDisabled) || $elements.findParent(mouseDownPhase.targetEl, canBeDisabled)
+
+        // Then check if parent is indeed disabled
+        if (parentThatCanBeDisabled !== null && $elements.isDisabled(parentThatCanBeDisabled)) {
```
Since `isDisabled` needs to be a `JQuery` object, we need to wrap the `HTMLElement` returned from `findParentThatCanBeDisabled`
```suggestion
        if (parentThatCanBeDisabled !== null && $elements.isDisabled($(parentThatCanBeDisabled))) {
```

AtofStryker commented on cli/CHANGELOG.md:
```diff
@@ -33,6 +33,7 @@ _Released 1/16/2024_
 - Fixed an issue where visiting an `http://` site would result in an infinite reload/redirect loop in Chrome 114+. Fixes [#25891](https://github.com/cypress-io/cypress/issues/25891).
 - Fixed an issue where requests made from extra tabs do not include their original headers. Fixes [#28641](https://github.com/cypress-io/cypress/issues/28641).
 - Fixed an issue where `cy.wait()` would sometimes throw an error reading a property of undefined when returning responses. Fixes [#28233](https://github.com/cypress-io/cypress/issues/28233).
+- Fixed an issue where `.click()` commands on children of disabled elemenets would still produce "click" events -- even without {force: true}. Fixes [#28788](https://github.com/cypress-io/cypress/issues/28788).
```
```suggestion
- Fixed an issue where `.click()` commands on children of disabled elements would still produce "click" events -- even without `{ force: true }`. Fixes [#28788](https://github.com/cypress-io/cypress/issues/28788).
```

AtofStryker replied:
this entry will need to be moved to `13.6.5`, which needs to be added. If you want me to take care of that I can @TheoAnastasiadis!

AtofStryker reviewed (changes requested):
Hi @TheoAnastasiadis! Thank you for the contribution PR. I had some issues getting the added code in `mouse.ts` to work correctly. Were you able to run the code changes against the cypress test or was this a best guess effort? Either way I think we can get this running with the suggestions below.

TheoAnastasiadis committed:
Update packages/driver/src/cy/mouse.ts

Co-authored-by: Bill Glesias <bglesias@gmail.com>

TheoAnastasiadis committed:
Update packages/driver/cypress/e2e/commands/actions/click.cy.js

Co-authored-by: Bill Glesias <bglesias@gmail.com>

TheoAnastasiadis committed:
Update packages/driver/src/cy/mouse.ts

Co-authored-by: Bill Glesias <bglesias@gmail.com>

TheoAnastasiadis committed:
Update packages/driver/src/cy/mouse.ts

Co-authored-by: Bill Glesias <bglesias@gmail.com>

TheoAnastasiadis committed:
update mouse.ts -- unexpected token

TheoAnastasiadis committed:
docs: moved entry to 13.6.5

TheoAnastasiadis committed:
Merge branch 'develop' into disabled-click

TheoAnastasiadis commented:
> Hi @TheoAnastasiadis! Thank you for the contribution PR. I had some issues getting the added code in `mouse.ts` to work correctly. Were you able to run the code changes against the cypress test or was this a best guess effort? Either way I think we can get this running with the suggestions below.

this was embarrassing 😳 is ok now, what should the release date of 13.6.5 be?

jennifer-shehane commented:
@TheoAnastasiadis 2 weeks from the previous release is the target, so 2/13/2024

TheoAnastasiadis committed:
Update CHANGELOG.md

TheoAnastasiadis committed:
Update CHANGELOG.md

AtofStryker commented:
> > Hi @TheoAnastasiadis! Thank you for the contribution PR. I had some issues getting the added code in `mouse.ts` to work correctly. Were you able to run the code changes against the cypress test or was this a best guess effort? Either way I think we can get this running with the suggestions below.
> 
> this was embarrassing 😳 is ok now, what should the release date of 13.6.5 be?

no worries @TheoAnastasiadis! We've all been there before 🙂 . CI should be running against the PR now.

AtofStryker commented on cli/CHANGELOG.md:
```diff
@@ -1,8 +1,20 @@
 
 _Released 1/30/2024_
 
+**Bugfixes:**
+
+- Fixed an issue with capturing assets for Test Replay when service workers are registered in Cypress support files. This issue would cause styles to not render properly in Test Replay. Fixes [#28747](https://github.com/cypress-io/cypress/issues/28747).
+
```
```suggestion
```

AtofStryker commented on cli/CHANGELOG.md:
```diff
@@ -1,8 +1,20 @@
+
+**Bugfixes:**
+
+Fixed an issue where `.click()` commands on children of disabled elements would still produce "click" events -- even without `{ force: true }`. Fixes [#28788](https://github.com/cypress-io/cypress/issues/28788).
```
```suggestion
- Fixed an issue where `.click()` commands on children of disabled elements would still produce "click" events -- even without `{ force: true }`. Fixes [#28788](https://github.com/cypress-io/cypress/issues/28788).
```

AtofStryker reviewed (commented):
Besides the changelog edits I think we are good to go! Going to wait for CI test feedback then will re review

AtofStryker commented on packages/driver/cypress/e2e/commands/actions/click.cy.js:
```diff
@@ -546,6 +546,27 @@ describe('src/cy/commands/actions/click', () => {
+
+      cy.get('button:first span').click()
+
+      cy.getAll('btn', 'mouseenter mousedown mouseup').each(shouldBeCalled)
```
Looks like this test is failing in CI on this line @TheoAnastasiadis https://app.circleci.com/pipelines/github/cypress-io/cypress/59642/workflows/13231e3d-4919-4cf4-87e0-59f684beb5d4/jobs/2482293/tests#failed-test-0. If I get some time I should be able to see what is going on

AtofStryker replied:
OK looks like chrome fires the mouse enter event on the parent element as well, but not in electron. We should be able to do something like this, or remove the assertion
```suggestion
      if (Cypress.browser.name === 'chrome') {
          cy.getAll('btn', 'mouseenter mousedown mouseup').each(shouldBeCalled)
      }
```

jennifer-shehane committed:
Merge branch 'develop' into disabled-click

TheoAnastasiadis committed:
Update packages/driver/cypress/e2e/commands/actions/click.cy.js

Co-authored-by: Bill Glesias <bglesias@gmail.com>

TheoAnastasiadis committed:
Update cli/CHANGELOG.md

Co-authored-by: Bill Glesias <bglesias@gmail.com>

TheoAnastasiadis committed:
Merge branch 'develop' into disabled-click

TheoAnastasiadis committed:
lint: fixed 2 linting errors

565:1  error  Expected indentation of 8 spaces but found 10  indent
567:7  error  Expected blank line before this statement      padding-line-between-statements

TheoAnastasiadis committed:
style: removed trailing space

567:1  error  Trailing spaces not allowed  no-trailing-spaces

jennifer-shehane committed:
Merge branch 'develop' into disabled-click

TheoAnastasiadis committed:
Merge branch 'develop' into disabled-click

TheoAnastasiadis committed:
Merge branch 'develop' into disabled-click

cacieprins committed:
Update CHANGELOG.md
